staging/src/k8s.io/apiserver/pkg/server/filters/goaway_test.go

```diff
@@ -0,0 +1,307 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"crypto/tls"
+	"io"
+	"math/rand"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"golang.org/x/net/http2"
+)
+
+func TestProbabilisticGoawayDecider(t *testing.T) {
+	cases := []struct {
+		name         string
+		chance       float64
+		nextFn       func(chance float64) func() float64
+		expectGOAWAY bool
+	}{
+		{
+			name:   "always not GOAWAY",
+			chance: 0,
+			nextFn: func(chance float64) func() float64 {
+				return rand.Float64
+			},
+			expectGOAWAY: false,
+		},
+		{
+			name:   "always GOAWAY",
+			chance: 1,
+			nextFn: func(chance float64) func() float64 {
+				return rand.Float64
+			},
+			expectGOAWAY: true,
+		},
+		{
+			name:   "hit GOAWAY",
+			chance: rand.Float64(),
+			nextFn: func(chance float64) func() float64 {
+				return func() float64 {
+					return chance - 0.01
```

Comment:
this could return negative numbers, which rand.Float64() cannot. It's probably OK but just mentioning.

Reply:
done